hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java

```diff
@@ -110,16 +112,18 @@ public void setTargetAddress(Address targetAddress) {
     }
 
     @Override
-    public void blacklist(Address address, boolean permanent) {
-        Boolean prev = blacklistedAddresses.putIfAbsent(address, permanent);
+    public void blacklist(Address address, boolean permanent, Throwable cause) {
+        Tuple2<Boolean, Throwable> prev = blacklistedAddresses.putIfAbsent(address, Tuple2.tuple2(permanent, cause));
         if (prev == null) {
             logger.info(address + " is " + (permanent ? "permanently " : "") + "added to the blacklist.");
         }
     }
 
     @Override
     public boolean unblacklist(Address address) {
-        if (blacklistedAddresses.remove(address, Boolean.FALSE)) {
+        if (blacklistedAddresses.containsKey(address)
+                && Objects.equals(blacklistedAddresses.get(address).f0(), Boolean.FALSE)) {
```

Comment:
This part is already removed but to answer your question, this is because of `remove` implementation. It is equivalent to:
```
 if (map.containsKey(key) && Objects.equals(map.get(key), value)) {
   map.remove(key);
   return true;
 } else
   return false;
```
But in my case I needed not `map.get(key)` but `map.get(key).f0()` hence I wrote like this. Technically it can throw NPE if `unblacklist` is called upon an address which is not blacklisted if it isn't written like this. Whether this could happen or not I didn't look at extensively but I just replaced current logic with equivalent newer one.